Spec: Clarify variant type classification and primitive type scoping (#16836)
apache/iceberg: format/spec.md

A **`variant`** is a value that stores semi-structured data. The structure and data types in a variant are not necessarily consistent across rows in a table or data file. The variant type and binary encoding are defined in the [Parquet project](https://github.com/apache/parquet-format/blob/master/VariantEncoding.md), with support currently available for V1. Support for Variant is added in Iceberg v3.

> The `variant` type is a semi-structured type. It is neither a primitive type nor a nested type.

Reviewer: Could follow the above paragraph and tighten a bit:
```
As a semi-structured type, `variant` is neither a primitive type nor a nested type.
```

Author: Thanks for the wording suggestion, applied your way :-)

> The `variant` type is a semi-structured type. It is neither a primitive type nor a nested type.

Author: ```suggestion
As a semi-structured type, `variant` is neither a primitive type nor a nested type.
```

| Transform name    | Description                                                  | Source types                                                                                              | Result type |
|-------------------|--------------------------------------------------------------|-----------------------------------------------------------------------------------------------------------|-------------|

> | **`identity`**    | Source value, unmodified                                     | Any primitive type except for `geometry` and `geography`                                                  | Source type |

Reviewer 2: let's align this with expressions-spec.md, which uses: Any primitive except for `geometry`, `geography`, and `variant`

Reviewer 3: should we include or exclude variant? 

Dan suggested we can exclude variant
> Including Variant in the exclusion does create some dissonance because it's already covered by the primitive scoping.

Reviewer 2: I'm +1 for aligning the wording across files

> | **`identity`**    | Source value, unmodified                                     | Any primitive except for `geometry` and `geography`. Nested types (`struct`, `list`, and `map`) and `variant` are not supported. | Source type |

Reviewer 4: ```suggestion
| **`identity`**    | Source value, unmodified                                     | Any primitive except for `geometry` and `geography` | Source type |
```

|-------------|--------|---------------------------|---------------------------|-----------------------------------------------|-------------|
| _optional_  | 1      | `lower_bound`             | Field type or `geo_lower` | all primitives or `variant`                   | Lower bound stored as the field's type, or `geo_lower` for geo types |
| _optional_  | 2      | `upper_bound`             | Field type or `geo_upper` | all primitives or `variant`                   | Upper bound stored as the field's type, or `geo_upper` for geo types |

> | _optional_  | 3      | `tight_bounds`            | `boolean`                 | all primitives except for `geometry` and `geography` | When true, `lower_bound` and `upper_bound` must be equal to the min and max values |

Reviewer 2: I don't see why we're changing this part of the content stats as this isn't related to identity transforms

Reviewer 2: nvm, after re-reading this is fine
